Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp

```diff
@@ -0,0 +1,271 @@
+    yshift_FFTtoCell = k_space.getSpectralShiftFactor(dm, 1,
+                                    ShiftType::TransformToCellCentered);
+    zshift_FFTfromCell = k_space.getSpectralShiftFactor(dm, 2,
+                                    ShiftType::TransformFromCellCentered);
+    zshift_FFTtoCell = k_space.getSpectralShiftFactor(dm, 2,
+                                    ShiftType::TransformToCellCentered);
+#else
+    zshift_FFTfromCell = k_space.getSpectralShiftFactor(dm, 1,
+                                    ShiftType::TransformFromCellCentered);
+    zshift_FFTtoCell = k_space.getSpectralShiftFactor(dm, 1,
+                                    ShiftType::TransformToCellCentered);
+#endif
+
+    // Allocate and initialize the FFT plans
+    forward_plan = FFTplans(spectralspace_ba, dm);
+    backward_plan = FFTplans(spectralspace_ba, dm);
+    // Loop over boxes and allocate the corresponding plan
+    // for each box owned by the local MPI proc
+    for ( MFIter mfi(spectralspace_ba, dm); mfi.isValid(); ++mfi ){
+        Box bx = spectralspace_ba[mfi];
+#ifdef AMREX_USE_GPU
+        // Add cuFFT-specific code
+#else
+        // Create FFTW plans
+        forward_plan[mfi] =
+            // Swap dimensions: AMReX FAB are Fortran-order but FFTW is C-order
+#if (AMREX_SPACEDIM == 3)
+            fftw_plan_dft_3d( bx.length(2), bx.length(1), bx.length(0),
+#else
+            fftw_plan_dft_2d( bx.length(1), bx.length(0),
+#endif
+            reinterpret_cast<fftw_complex*>( tmpRealField[mfi].dataPtr() ),
+            reinterpret_cast<fftw_complex*>( tmpSpectralField[mfi].dataPtr() ),
+            FFTW_FORWARD, FFTW_ESTIMATE );
+        backward_plan[mfi] =
+            // Swap dimensions: AMReX FAB are Fortran-order but FFTW is C-order
+#if (AMREX_SPACEDIM == 3)
+            fftw_plan_dft_3d( bx.length(2), bx.length(1), bx.length(0),
+#else
+            fftw_plan_dft_2d( bx.length(1), bx.length(0),
+#endif
+            reinterpret_cast<fftw_complex*>( tmpSpectralField[mfi].dataPtr() ),
+            reinterpret_cast<fftw_complex*>( tmpRealField[mfi].dataPtr() ),
+            FFTW_BACKWARD, FFTW_ESTIMATE );
+#endif
+    }
+}
+
+
+SpectralFieldData::~SpectralFieldData()
+{
+    if (tmpRealField.size() > 0){
+        for ( MFIter mfi(tmpRealField); mfi.isValid(); ++mfi ){
+#ifdef AMREX_USE_GPU
+            // Add cuFFT-specific code
+#else
+            // Destroy FFTW plans
+            fftw_destroy_plan( forward_plan[mfi] );
+            fftw_destroy_plan( backward_plan[mfi] );
+#endif
+        }
+    }
+}
+
+/* \brief Transform the component `i_comp` of MultiFab `mf`
+ *  to spectral space, and store the corresponding result internally
+ *  (in the spectral field specified by `field_index`) */
+void
+SpectralFieldData::ForwardTransform( const MultiFab& mf,
+                                     const int field_index,
+                                     const int i_comp )
+{
```

Comment:
May be worth it to add `BL_PROFILE("SpectralFieldData::ForwardTransform()");` here and in `BackwardTransform`?

Reply:
Yes, definitely! Any preference as to whether to add this around the call in `SpectralSolver`, or here?

Reply:
Right, probably better to put it in `SpectralSolver` for future developments